walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/TimeRange.java

```diff
@@ -183,4 +184,35 @@ public static TimeRange fromStartEnd(int start, int end, boolean inclusive) {
   public static TimeRange fromStartDuration(int start, int duration) {
     return new TimeRange(start, duration);
   }
+
+  /**
+   * Uses a branchless binary search to find the TimeRange with the biggest starting point,
+   * that is <= to the starting point of @param target. On an empty @param collection,
+   * returns 0. On a null @param collection or @param target, returns -1.
+   * It searches the index of the @param target bit by bit, starting with the MSB and going
+   * down to the LSB of the index.
+   */
+  public static int lowerBound(final List<TimeRange> collection, final TimeRange target) {
+    if (collection == null || target == null) {
+      return -1;
+    }
+
+    int lowerBoundPosition = 0;
+
+    for (int pace = maxPossiblePace(collection.size()); pace > 0; pace >>= 1) {
+      if (pace + lowerBoundPosition < collection.size()) {
+        lowerBoundPosition += pace * (collection.get(lowerBoundPosition + pace).start() <=
+                                      target.start() ? 1 : 0);
+      }
+    }
+
+    return lowerBoundPosition;
+  }
+
+  /**
+   * Finds the biggest power of 2 closest to the @param length with the formula 2 ^ [log_2(n)].
+   */
+  private static int maxPossiblePace(final int length) {
+    return (int)Math.pow(2, (int)(Math.log(length) / Math.log(2)));  
```

Comment:
that's _a bit_ too much conversion and floating point conversion for my taste.
Would [Integer.highestOneBit](https://www.geeksforgeeks.org/integer-highestonebit-method-in-java/) be for this?

Reply:
Yes! Thank you for finding this method, I didn't know it existed.

P.S. Really nice pun there 😀